### DISCUSSION TITLE
Minimum time minimum control effort solution

### DISCUSSION THREAD

**Discussion by iparaj** [casadi/casadi#3582] [Q&A]
Can I optimize on the time T in the OCP at the same time optimizing on L (The Lagrangian term) using CasADi?
![Screenshot 2024-02-09 172226](https://github.com/casadi/casadi/assets/74721068/d9911c77-b801-488c-a872-7d0f1c60bc00)

**Comment by brunoml5** [comment]
Yes. Take a look on the `race_car.py` in the example pack.

Also, there other ways of writing your ODE to consider variable time. Another option is to rewrite your ODE so that time is another state.

**Reply by iparaj** [reply]
The race_car.py example solves only the minimum time OCP. I am looking for minimum time minimum control effort OCP, much like the free final time fixed final state standard OCP. Can you throw me some resources to write time as a state or can you help me out on this?

**Reply by brunoml5** [reply]
Do you want to minimize control effort? Just add it to your objective.

I tried looking for a reference and could not easily find it. To write time as a state, you need to do a change of variable. Take a look at this [post](https://github.com/casadi/casadi/discussions/3249#discussioncomment-6253354) from Joel.

Let's say you want to change your system from `dx/dt` to `dx/dz`. Then, you write your system as:
```
dx_dz = Ts*dx_dt
dt_dz = Ts
```
Then you integrate both `x` and `t` from 0 to 1 where `Ts` is the sampling time. Depending on how you write your OCP and your integration method, `Ts` can be a parameter or a optimization variable.
